### ISSUE TITLE
Client-side code generation

### ISSUE THREAD

**Issue by kobylynskyi** [kobylynskyi/graphql-java-codegen#37]
**Is your feature request related to a problem? Please describe.**
Currently, the main purpose of the plugin is to generate server-side code (POJOs and interfaces). 
The idea is to add support of client-side classes generation. This will allow users to build a GraphQL request.

**Describe the solution you'd like**
Schema example:
```graphql
type Mutation {
    newBike(bikeInput: BikeInput!): Bike
}

input BikeInput {
    type: BikeType!
    brand: String!
    size: String!
    price: BigDecimal
}

type Bike {
    id: ID
    type: BikeType
    brand: String
    size: String
    price: BigDecimal
    addedDateTime: DateTime
}
```

Generated class:

```java
public class NewBikeMutationRequest {
    private BikeInput bikeInput;
    //getters/setters
}

public class BikeInputTO {
    private BikeTypeTO type;
    private String brand;
    private String size;
    private String price;

    public BikeInputTO() { }

    //getters/setters
```

The end-user will then use these classes to build the request and **response projection** (open question for now). So that it will be serialized to the following request string:

```graphql
mutation newBike {
  newBike(bike: {
    type: ROAD,
    brand: "LOL",
    size: "Big",
    year: 2019,
    price: "123.45"
  }) {
    id
    addedDateTime
  }
}
```

**Comment by kobylynskyi** [comment]
@joffrey-bion 
I was not able to find a solution for client-side code generation in Java.

The idea of this feature would be:
1. Define the projection of the response object (which fields to fetch)
2. Properly serialize the request.

So the code on the client-side will look like:
```java
BikeInputTO input = new BikeInputTO(...);
NewBikeMutationRequest request = new NewBikeMutationRequest(input)
                .responseProjection(new BikeProjection()
                        .id().addedDateTime());
Bike bikeResponse = restTemplate.postForObject("https://localhost:8080", request, Bike.class);
```
(the DSL is still discussable)

**Comment by joffrey-bion** [comment]
> I was not able to find a solution for client-side code generation in Java.

Yes, I was actually mentioning code generators in general.
The one I heard about was for Swift I think.

> The idea of this feature would be:
> 1. Define the projection of the response object (which fields to fetch)
> 2. Properly serialize the request.

Oh OK, so you want to build client queries in Java directly?
That could be a way too, I just felt like the idea of this library was to go schema-first, so I expected a query-first solution here. But that implies parsing queries. I don't know if the library you're using can do this.

**Comment by kobylynskyi** [comment]
> Oh OK, so you want to build client queries in Java directly?
Yes. But the approach will still be **schema-first** because the client will use schema to generate the request and response classes.
The main problem that I am facing in one of my projects is to serialize the request. Because GraphQL request is not a JSON and I can't find a proper GraphQL request serialization in java.

**Comment by JustinBlandford** [comment]
Hi @kobylynskyi that's exactly what I was thinking about and thanks for raising this feature,  I initially thought that the graphql-codegen-maven-plugin plugin did generate client code as it generates classes that do seem to map to the response class :-), A new feature the way you have illustrated above would be very handy.  Ultimately, all I'd like to is to be able to run a GraphQL query by building a request in Java and interrogate the response (similar to the swagger-codegen-maven-plugin).  Schema first is critical as it dynamically creates the classes when changes to the schema are made as you have re-iterated above.

Do you have any rough idea of when such a feature may be available to use?

Thanks again,
Justin

**Comment by kobylynskyi** [comment]
@JustinBlandford
Thanks for your feedback. I am glad that you found this helpful.
I am targeting to release this feature by the end of this week.

**Comment by JustinBlandford** [comment]
@kobylynskyi that's great, thanks

**Comment by kobylynskyi** [comment]
@JustinBlandford 
This feature is available in plugin version 1.6.0.
* Sample plugin configuration: https://github.com/kobylynskyi/graphql-java-codegen/blob/master/plugins/maven/example-client/pom.xml
* Sample GraphQL request using generated classes: https://github.com/kobylynskyi/graphql-java-codegen/blob/master/plugins/maven/example-client/src/main/java/io/github/kobylynskyi/order/external/ProductServiceGraphQLClient.java#L37

Please feel free to open an issue in this project or feature request if you have any suggestions/ideas.
Thanks!
